rocks-lib/src/upload/mod.rs

```diff
@@ -0,0 +1,236 @@
+
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub enum UploadError {
+    Lua(#[from] mlua::Error),
+    Request(#[from] reqwest::Error),
+    RockCheck(#[from] RockCheckError),
+    #[error("rock already exists on server: {0}")]
+    RockExists(String),
+    #[error("unable to read rockspec: {0}")]
+    RockspecRead(#[from] std::io::Error),
+    Signature(#[from] gpgme::Error),
+    ToolCheck(#[from] ToolCheckError),
+    UserCheck(#[from] UserCheckError),
+}
+
+pub struct ApiKey(String);
+
+#[derive(Error, Debug)]
+#[error("no API key provided! Please set the $ROCKS_API_KEY variable")]
+pub struct ApiKeyUnspecified;
+
+impl ApiKey {
+    /// Retrieves the rocks API key from the `$ROCKS_API_KEY` environment
+    /// variable and seals it in this struct.
+    pub fn new() -> Result<Self, ApiKeyUnspecified> {
+        Ok(Self(
+            env::var("ROCKS_API_KEY").map_err(|_| ApiKeyUnspecified)?,
+        ))
+    }
+
+    /// Creates an API key from a String.
+    ///
+    /// # Safety
+    ///
+    /// This struct is designed to be sealed without a [`Display`](std::fmt::Display) implementation
+    /// so that it can never accidentally be printed.
+    ///
+    /// Ensure that you do not do anything else with the API key string prior to sealing it in this
+    /// struct.
+    pub unsafe fn from(str: String) -> Self {
+        Self(str)
+    }
+
+    /// Retrieves the underlying API key as a [`String`].
+    ///
+    /// # Safety
+    ///
+    /// Strings may accidentally be printed as part of its [`Display`](std::fmt::Display)
+    /// implementation. Ensure that you never pass this variable somewhere it may be displayed.
+    pub unsafe fn get(&self) -> &String {
+        &self.0
+    }
+}
+
+pub async fn upload_from_project(
+    project: &Project,
+    api_key: &ApiKey,
+    config: &Config,
+) -> Result<(), UploadError> {
+    let client = Client::new();
+
+    helpers::ensure_tool_version(&client, config.server()).await?;
+    helpers::ensure_user_exists(&client, api_key, config.server()).await?;
+
+    let rockspec = project.rockspec();
+
+    if helpers::rock_exists(
+        &client,
+        api_key,
+        &rockspec.package,
+        &rockspec.version,
+        config.server(),
+    )
+    .await?
+    {
+        return Err(UploadError::RockExists(config.server().clone()));
+    }
+
+    let rockspec_content = std::fs::read_to_string(project.root().join("project.rockspec"))?;
+
+    let signed = match Context::from_protocol(Protocol::OpenPgp) {
+        Ok(mut ctx) => {
+            ctx.set_armor(true);
+
+            let mut signature = Data::new()?;
+
+            match ctx.sign_detached(rockspec_content.clone(), &mut signature) {
+                Ok(_) => {
+                    let mut signature_str = String::new();
+                    signature.read_to_string(&mut signature_str)?;
+
+                    Some(signature_str)
+                }
+                // TODO: Issue a warning here that signing wasn't successful.
```

Comment:
suggestion: Le'ts not issue a warning. Instead, let's default to returning an `Err`, suggesting to pass in a `--skip-sign` flag.
I think it's better if people have to explicitly skip signing.

Reply:
Done. For consistency's sake I made it a `--sign-protocol` flag, allowing the user to select their protocol of choice. Whenever GPG issues an error, the user also receives a hint saying they can pass `--sign-protocol none` to skip the signing step :)